### DISCUSSION TITLE
middleware vs running after connection

### DISCUSSION THREAD

**Discussion by TriStarGod** [socketio/socket.io#4249] [Q&A] [answered]
Hi,
I'm working with socket.io for the first time to setup auth.
Whats the difference between using a middleware vs running a function after connection?

Ref: https://socket.io/docs/v4/middlewares/

```
// middleware
io.use((s, next) => {
 if (isAuthed(s.request)) {
  next();
 } else {
  next(new Error("invalid"));
 }
});
```

vs

```
// after connection
io.of("/").on("connection", (s) => {
 if (!isAuthed(s.request)) throw new Error("invalid");
 ...
});
```

I asked this question on [stackoverflow](https://stackoverflow.com/questions/70658058/socket-io-middleware-vs-running-after-connection) but no one answered.

**Comment by darrachequesne** [comment, marked as answer]
Hi! Technically, there is no difference, the underlying connection (HTTP long-polling or WebSocket) is already established.

That being said, middlewares are the best place to check if the user is authenticated/has sufficient permissions (before the connection at the Socket.IO level). I guess it makes more sense from a client point of view:

- middleware: the client receives a `connect_error` event upon failure
- in the `connection` handler:  the client first receives a `connect` event and then a `disconnect` event

**Reply by TriStarGod** [reply]
Thanks! That makes a difference to my frontend setup.
